### ISSUE TITLE
Add tarteaucitron.js lib integration

### ISSUE THREAD

**Issue by thomas-blackbird** [opengento/magento2-gdpr#9]
Should support https://github.com/AmauriC/tarteaucitron.js

**Comment by thomas-kl1** [comment]
Each 13 months, customer must give his consentent of personal data processing (cookies).

**Comment by SomeGeek** [comment]
Important to note: As of 2.3, a reCaptcha module is included by default. In most situations, this will be turned on, at least for the contact form. 

However, as this collects _lots_ of personal data, this can't be considered a functional cookie.

So, to comply with the law, people need to opt-in for this. But if they don't it is not an option to present them with a unprotected contact form either, that would be a bad idea.

I see only really 2 ways around this:

- Use a honeypot. Not a favorite option as it's quite easy to work around it for bots. And if it's not today, it will be tomorrow: spammers innovate, just like us, unfortunately.
- Prevent submitting the form without reCatpcha consent. Do this by only processing the form if reCaptcha cookies are sent with the form.

The latter one requires fine-grained cookie control, thus having multiple (configurable) groups. Like Marketing, Analytics and reCaptcha.

**Comment by SomeGeek** [comment]
Another note: We also need to deal with other third-party cookie-enabled extensions which are included by default. Google Analytics and Google Adwords are two of them, which I suspect are used a lot.

Magento 2 has a internal "cookie restriction mode", which gets checked before those are loaded. We could extend that to also look at our cookie settings.

**Comment by thomas-kl1** [comment]
It might be necessary to achieve this in a specific module, as it can quickly became large. We definitly need to check what is involved in Magento core.

**Comment by ioweb-gr** [comment]
I would like to point out for tarteaucitron that it's actually a very good solution regarding the fact that it allows you out of the box to run a script only after the user gives his consent explicitly. However due to it's legacy javascript code, it's very hard to ensure that it's properly loaded before you load your custom scripts.

It's impossible to handle the dependency with requirejs and keep it modular so often it will give issues with uninitialized window objects that google tags need. For example remarketing code in product page/cart etc is not certain that will execute correctly unless you tie it in the custom script code. 

It's also loading it's own file for services via the main tarteaucitron.js file so merging usually breaks it. You'd have to load it from a CDN

Basically it's very hard to integrate in asynchronous ways.

**Comment by thomas-kl1** [comment]
Interesting.. thank you for your feedback, have you proof of concept with Magento 2 or integration tool to achieve this in mind?

**Comment by ioweb-gr** [comment]
I've been experimenting with this for a while because I haven't seen a single module offering proper cookie consent integration for Magento 2 but every single one has limitations. You'd be amazed to see that even paid solutions don't adhere to the regulation regarding cookies. For example some just unset the cookies instead of preventing their storage on the user's device in the first place before consent is given. Only tarte was 100% compliant but unreliable for Magento 2.

I've tried with quantcast GDPR too but it relies on external service so I avoided that too. Seems similar to osano mentioned earlier.

I am currently focusing on this library

https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework

I haven't tried it yet since it needs studying but the main idea is that the framework provides a list of purposes and the ability to check if consent is given for them.

So you can wrap scripts and inline scripts with a check for which purpose you need consent for. If consent is given then the code executes

**Comment by owebia** [comment]
@thomas-kl1 anything new on this?

I'm using recaptcha and Google Analytics (Google Tag Manager).
I also use tarteaucitron.js.

Google Analytics can easily be configured with tarteaucitron.js (instead of using the built-in functionality).

What is interesting with tarteaucitron.js is that you can put a placeholder where cookies are required.
It allows the visitor to easily accept additional cookies when he wants to complete some actions (like using the contact form).

![Nous-contacter](https://user-images.githubusercontent.com/604741/124315225-ebe04e80-db73-11eb-811d-213e52d5135d.png)

**Comment by owebia** [comment]
I created a proof of concept on delaying the loading of recaptcha to the cookie consent.

Opengento/Gdpr/view/frontend/requirejs-config.js
```js
 var config = {
    config: {
        mixins: {
            'Magento_ReCaptchaFrontendUi/js/reCaptchaScriptLoader': {
                'Opengento_Gdpr/js/reCaptchaScriptLoader-mixin': true
            }
        }
    }
};
```

Opengento/Gdpr/view/frontend/web/js/reCaptchaScriptLoader-mixin.js
```js
 define([
    'jquery',
    'mage/utils/wrapper',
    'mage/cookies'
], function (
    $,
    wrapper
) {
    'use strict';

    return function (reCaptchaScriptLoader) {
        reCaptchaScriptLoader.addReCaptchaScriptTag = wrapper.wrapSuper(
            reCaptchaScriptLoader.addReCaptchaScriptTag,
            function () {
                return $.cookie('cookies-policy') === '1' ? this._super() : null;
            }
        );
        if ($.cookie('cookies-policy') !== '1') {
            var interval = setInterval(function () {
                if ($.cookie('cookies-policy') === '1') {
                    reCaptchaScriptLoader.addReCaptchaScriptTag();
                    clearInterval(interval);
                }
            }, 3000);
        }
        return reCaptchaScriptLoader;
    };
});
```

Tell me if you want me to create a pull request

**Comment by owebia** [comment]
I looked at how Google Analytics could be delayed and it seems that Magento already implemented something with their cookie notice.

https://github.com/magento/magento2/blob/33242e4b19cf207d7b73f7791ef894b48bb41f8a/app/code/Magento/GoogleAnalytics/view/frontend/web/js/google-analytics.js#L21-L33

Here is the name of the cookie:
https://github.com/magento/magento2/blob/2.4.2/app/code/Magento/Cookie/view/frontend/templates/html/notices.phtml#L40
https://github.com/magento/magento2/blob/33242e4b19cf207d7b73f7791ef894b48bb41f8a/app/code/Magento/Cookie/Helper/Cookie.php#L18

An event is also triggered (but not used) when the user allows cookies:
https://github.com/magento/magento2/blob/33242e4b19cf207d7b73f7791ef894b48bb41f8a/app/code/Magento/Cookie/view/frontend/web/js/notices.js#L33

Should this module override the default behavior of cookie notice or should it be completely separated?

**Comment by owebia** [comment]
I created a proof of concept on delaying the loading of google analytics to the cookie consent.

Opengento/Gdpr/view/frontend/requirejs-config.js
```js
 var config = {
    config: {
        mixins: {
            'Magento_ReCaptchaFrontendUi/js/reCaptchaScriptLoader': {
                'Opengento_Gdpr/js/reCaptchaScriptLoader-mixin': true
            },
            'Magento_GoogleAnalytics/js/google-analytics': {
                'Opengento_Gdpr/js/google-analytics-mixin': true
            }
        }
    }
};
```

Opengento/Gdpr/view/frontend/web/js/google-analytics-mixin.js
```js
define([
    'jquery',
    'mage/cookies'
], function (
    $
) {
    'use strict';

    return function (googleAnalytics) {
        return function (config) {
            if ($.cookie('cookies-policy') === '1') {
                googleAnalytics(config);
            } else {
                var interval = setInterval(function () {
                    if ($.cookie('cookies-policy') === '1') {
                        googleAnalytics(config);
                        clearInterval(interval);
                    }
                }, 3000);
            }
        };
    };
});
```

**Comment by thomas-kl1** [comment]
Hi @owebia that is really interesting. I've never took attention to the cookie restriction feature available in Magento. It already does better than the "cookie disclosure" added by this module. Actually it could great if our module could extends the native feature, so the consumer is allowed to use an alternative (such as tarteaucitron).

So basically, the current "cookie disclosure" feature, if it's not improved, I'll remove it in favor of the cookie restriction from Magento. So yes, let's extends it, maybe we could add setting in the admin in order to tell which services must subscribes to `config.isCookieRestrictionModeEnabled`

What's your thoughts?

**Comment by owebia** [comment]
I agree it would be cleaner to override the default cookie restriction feature available in Magento (same cookie name, use triggered event, etc.).

But some challenges remain: tarteaucitron allows the user to choose the cookies he allows or refuses (he can allow recaptcha and refuse Google Analytics for example).  
The default behavior of Magento is only an on/off toggle.  
If we want to comply, we need to allow the user to choose which cookie he allows/blocks.
We also need to store and document the consent.

See https://gdpr.eu/cookies/, title "Cookie compliance".

**Comment by owebia** [comment]
An important question to answer: is consentement required for using reCAPTCHA and Google Analytics?

For France, the response is:
- __reCAPTCHA :__ yes
  https://www.cnil.fr/fr/application-stopcovid-la-cnil-tire-les-consequences-de-ses-controles
  https://www.legifrance.gouv.fr/cnil/id/CNILTEXT000042125452
  _« veiller, le cas échéant, à informer et recueillir le consentement des personnes concernées aux actions de lecture et d’écriture des informations présentes sur les terminaux de communication électronique par la société GOOGLE dans le cadre de la technologie reCaptcha, conformément aux dispositions de l’article 82 de la loi Informatique et Libertés. »_
- __Google Analytics :__ yes
  https://www.cnil.fr/fr/mesurer-la-frequentation-de-vos-sites-web-et-de-vos-applications
  _« Les offres de mesure d’audience n’entrent pas dans le périmètre de l’exemption notamment lorsque les fournisseurs indiquent réutiliser les données pour leur propre compte. C’est le cas notamment de plusieurs grandes offres de mesure d’audience disponibles sur le marché (voir notamment les politiques de confidentialité de Google Analytics, de Quantcast Analytics ou encore de Facebook Analytics). »_

**Comment by thomas-kl1** [comment]
@owebia ok so we definitely need a new entity to store consents. This could be achieved along #85  #86 
This is going to be an important amount of work :)
